compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala

```diff
@@ -750,7 +750,11 @@ object ProtoTypes {
         else
           newDepTypeVar(ref)
       mt.resultType.substParams(mt, mt.paramRefs.map(replacement))
-    else mt.resultType
+    else
+      // TODO(mbovel): document this.
+      mt.resultType match
+        case at@AppliedType(tycon: TypeRef, args) if defn.isCompiletimeAppliedType(tycon.symbol) => at.underlying
+        case _ => mt.resultType
```

Comment:
@smarter this is my proposition to fix the problem I mentioned in our last meeting.

Before this change :

```scala
def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: 6L = mult(2L, 3L) // 6L does not adds constraints on A or B, so there are inferred as Long and Long

// -- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:8:20 
// 8 |    val z: 6L = mult(2L, 3L)
//   |                ^^^^^^^^^^^^
//   |                Found:    Long * Long
//   |                Required: (6L : Long)
```

```scala
val m: Long = 42
def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: 2L * m.type // normalized to m.type * 2L
     = mult(2L, m) // A is inferred to be 2L and B to be m.type

// -- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:7:30 
// 7 |    val z: 2L * m.type = mult(2L, m)
//   |                              ^^
//   |                              Found:    (2L : Long)
//   |                              Required: (m : Long)
// 
// longer explanation available when compiling with `-explain`
// -- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:7:34 
// 7 |    val z: 2L * m.type = mult(2L, m)
//   |                                  ^
//   |                                  Found:    (m : Long)
//   |                                  Required: (2L : Long)

val m: Long = 42
def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: m.type * 2L // already in normal form
     = mult(2L, m) // A is inferred to be m.type and B to be 2L, so this works
```

After this change, singletons type parameters are never inferred for compile-time operations, so all of the 3 above cases fail with the same message:

```
Found:    Long * Long
Required: …
```

However, typing `mult` with `Singleton` bounds or with dependent parameters works fine before and after for all 3 cases:

```scala
def mult2(a: Long, b: Long) = (a * b).asInstanceOf[a.type * b.type]
def mult3[A <: Long & Singleton, B <: Long & Singleton](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
```

What do you think? Would that be a more coherent behavior? Or should we not change it?

In both cases tf-dotty would need to be updated to type operations as `mult2` or `mult3`. See https://github.com/MaximeKjaer/tf-dotty/blob/master/modules/compiletime/src/main/scala/io/kjaer/compiletime/dependent.scala#L9. I can't find of a simple solution that would not require changes in tf-dotty.

Reply:
> Would that be a more coherent behavior? 

No. It adds a special case in inference making it behave differently than for normal operations and the end result seems to only break more things, so I wouldn't do this.